### PR TITLE
Fix json trace/metrics unmarshaling for numbers

### PR DESCRIPTION
**Pull request by bogdandrutu**
Updates https://github.com/open-telemetry/opentelemetry-collector/issues/5312

Also this improves benchmarks since `ReadAny` was doing an allocation, see:

**Before:**
```
/private/var/folders/5c/5p_3jmvd6qx0rsvmb9j7c_s00000gn/T/GoLand/___BenchmarkJSONUnmarshal_in_go_opentelemetry_io_collector_pdata_ptrace.test -test.v -test.paniconexit0 -test.bench ^\QBenchmarkJSONUnmarshal\E$ -test.run ^$
goos: darwin
goarch: amd64
pkg: go.opentelemetry.io/collector/pdata/ptrace
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkJSONUnmarshal
BenchmarkJSONUnmarshal-16    	  334974	      3942 ns/op	    4669 B/op	     231 allocs/op
PASS
```

**After:**
```
/private/var/folders/5c/5p_3jmvd6qx0rsvmb9j7c_s00000gn/T/GoLand/___BenchmarkJSONUnmarshal_in_go_opentelemetry_io_collector_pdata_ptrace.test -test.v -test.paniconexit0 -test.bench ^\QBenchmarkJSONUnmarshal\E$ -test.run ^$
goos: darwin
goarch: amd64
pkg: go.opentelemetry.io/collector/pdata/ptrace
cpu: Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
BenchmarkJSONUnmarshal
BenchmarkJSONUnmarshal-16    	  374025	      3642 ns/op	    4412 B/op	     219 allocs/op
PASS

Signed-off-by: Bogdan <bogdandrutu@gmail.com>

### CODE REVIEW

**Comment by tigrannajaryan** [comment]
Github shows lots of line changes but it seems many of the changes are just moving the code without changing functionality. Can you summarize the changes (other than performance improvement)?

**Review by dmitryax — APPROVED** [review]

**Comment by bogdandrutu** [comment]
@tigrannajaryan good point, split the first simple PR https://github.com/open-telemetry/opentelemetry-collector/pull/5926 then will split one more time to have focus changes.

File: pdata/internal/json/enum.go [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+)
+
+// ReadEnumValue returns the enum integer value representation. Accepts both enum names and enum integer values.
+// See https://developers.google.com/protocol-buffers/docs/proto3#json.
+func ReadEnumValue(iter *jsoniter.Iterator, valueMap map[string]int32) int32 {
+	switch iter.WhatIsNext() {
+	case jsoniter.NumberValue:
+		return iter.ReadInt32()
+	case jsoniter.StringValue:
+		return valueMap[iter.ReadString()]
```
Comment (tigrannajaryan) [inline pdata/internal/json/enum.go:28]: When the string is not a valid enum name I think it should be an error instead of returning the zero value.

File: pdata/internal/json/number.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+)
+
+// ReadInt32 unmarshalls JSON data into an int32. Accepts both numbers and strings decimal.
+// See https://developers.google.com/protocol-buffers/docs/proto3#json.
+func ReadInt32(iter *jsoniter.Iterator) int32 {
+	switch iter.WhatIsNext() {
+	case jsoniter.NumberValue:
+		return iter.ReadInt32()
+	case jsoniter.StringValue:
+		val, err := strconv.ParseUint(iter.ReadString(), 10, 32)
```
Comment (tigrannajaryan) [inline pdata/internal/json/number.go:30]: To handle negative numbers correctly:
```suggestion
		val, err := strconv.ParseInt(iter.ReadString(), 10, 32)
```
Also would be good to add a test for negative numbers.

File: pdata/internal/json/enum.go (hunk 1 above)
Comment (bogdandrutu) [reply pdata/internal/json/enum.go:28]: That will not allow us to extend the values in a backwards compatible way.

File: pdata/internal/json/number.go (hunk 2 above)
Comment (bogdandrutu) [reply pdata/internal/json/number.go:30]: Will do.

File: pdata/internal/json/enum.go (hunk 1 above)
Comment (bogdandrutu) [reply pdata/internal/json/enum.go]: Moved to https://github.com/open-telemetry/opentelemetry-collector/pull/5928
